content/5_Plat/HLD.mdx

```diff
@@ -252,8 +252,329 @@ set problems where $\Theta(Q\sqrt N\log N)$ is intended ...
+	another node, then the edge to its parent is a light edge, making the node
+	the root of a heavy path. We then continue descending to the heavy children
+	until we reach a leaf. This guarantees that all heavy paths are continuous
+	segments of the array.
+	**/
+	HLD(int n, vector<int> &v, vector<vector<int>> &adj) {
+		this->n = n;
+		heavy.assign(n + 1, 0);
+		depth.assign(n + 1, 0);
+		root.assign(n + 1, 0);
+		pos.assign(n + 1, 0);
+		par.assign(n + 1, 0);
+		par[1] = 1;
+		dfs(1, 1, adj);
+
+		// Decomposition part begins here
+		vector<int> a;  // to store the values in the segment tree
+		for (int i = 1, curr_pos = 1; i <= n; ++i) {
+			if (i == 1 or heavy[par[i]] != i) {
+				for (int j = i; j; j = heavy[j]) {
+					root[j] = i;
+					pos[j] = curr_pos++;  // assign position in the segment tree
+					a.push_back(v[j]);
+				}
+			}
+		}
+		st.init(n, a);
+	}
+
+	// path queries
+	int query(int a, int b) {
+		int s = 0;
+		while (root[a] != root[b]) {
+			if (depth[root[a]] > depth[root[b]]) { swap(a, b); }
+			s = max(s, st.query(pos[root[b]], pos[b]));
+			b = par[root[b]];
+		}
+		if (depth[a] > depth[b]) { swap(a, b); }
+		s = max(s, st.query(pos[a], pos[b]));
+		return s;
+	}
+
+	// update queries
+	void update(int n, int v) { st.update(pos[n], v); }
+};
+```
+</CPPSection>
+</LanguageSection>
+
+<Info title = "Note">
+The above-mentioned template is about finding the maximum value along the path between $a$ and $b$. For other operations, we can modify our segment tree to find the solution to the queries involving the sum, minimum value between two nodes, etc.
+</Info>
+
+At first, we choose an arbitrary root, ususally the node with index $1$ (or $0$ if you use $0$-based indexing), and start a DFS over the tree to find, for each node, its depth and subtree size. After this, we will decompose the tree to eventually create one large segment tree where we can calculate the maximum of the node values over a range of the path, ordering the values of the nodes based on their positions in the path.
+
+The next important step is to find the Lowest Common Ancestor(LCA) of the two nodes for which we will move node $a$ and node $b$ up the tree until both of them are in the same heavy path. We also need to ensure that we are always moving the deeper node, for which we need to swap $a$ and $b$ in case depths of the roots of $a$ and $b$ are different. Once $a$ and $b$ are on the same chain, the higher pointer of $a$ and $b$ is the LCA.
+
+We can calculate maximum of the node values over the path from $a$ to $b$ by first calculating maximum over the ranges of heavy paths that are between $a$ and $b$, and then calculating the maximums of these values and those of the light edges, obtaining the result we seek.
+
+[This](https://github.com/freakin23/CSES/blob/main/Tree%20Algorithms/2134.cpp) is the code for "Path Queries II" using the above template.
+
+<FocusProblem problem = "sample3"/>
```

Comment:
follow format in: https://docs.google.com/document/d/1Dzj5uYNYNXe_E-_9-lSeX73654cGJk88PZ0kZCCmks0/edit

Reply:
is the issue related to the positioning of the focus problem?

Reply:
check it once